Fix creation of dataviews inferred with .NET types with sparse vectors

xadupre opened the pull request:
Fixes #586. `DataViewConstructionUtils`'s methods to create dataviews over .NET types will now have correctly inferred getters in the case of sparse vectors.

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/machinelearning?pullRequest=587) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/machinelearning?pullRequest=587) before we can accept your contribution.<br/><br/>:x: xadupre [sign now](https://cla.dotnetfoundation.org/dotnet/machinelearning?pullRequest=587)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/machinelearning?pullRequest=587) it.</sub>

TomFinley commented on test/Microsoft.ML.TestFramework/TestSparseDataView.cs:
```diff
@@ -0,0 +1,114 @@
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.ML.Runtime.Command;
```
I'm fairly certain not all of these are used.

Ivanidzo4ka commented on test/Microsoft.ML.TestFramework/TestSparseDataView.cs:
```diff
@@ -0,0 +1,114 @@
+                new SparseExample<T>() { X = new VBuffer<T> (5, 3, v1, new int[] { 0, 2, 4 }) },
+                new SparseExample<T>() { X = new VBuffer<T> (5, 3, v2, new int[] { 0, 1, 3 }) }
+            };
+            var host = new TlcEnvironment();
```
>var host = new TlcEnvironment(); [](start = 12, length = 32)

can you wrap it in using?

Ivanidzo4ka commented on test/Microsoft.ML.TestFramework/TestSparseDataView.cs:
```diff
@@ -0,0 +1,114 @@
+            GenericSparseDataView(new double[] { 1, 2, 3 }, new double[] { 1, 10, 100 });
+            GenericSparseDataView(new DvText[] { new DvText("a"), new DvText("b"), new DvText("c") },
+                                  new DvText[] { new DvText("aa"), new DvText("bb"), new DvText("cc") });
+            Done();
```
>Done(); [](start = 12, length = 7)

You don't have to call Done if you don't do any basefiles comparison

Ivanidzo4ka commented on test/Microsoft.ML.TestFramework/TestSparseDataView.cs:
```diff
@@ -0,0 +1,114 @@
+            Assert.True(n == 2);
+        }
+
+        [Fact]
```
>Fact [](start = 9, length = 4)

Thank you for fixing my mess!
Your code definitely do right thing, and it's nice to have tests, but I want to point one thing. Your tests never hit TypedCursor changes.
There is two conversions: one from some collection to IDataview through code in DataViewConstructionUltis, and from IDataVIew to Collection via TypedCursor and your tests cover only first case.

In order to invoke second case you need to do something like:
dataView.AsEnumerable<ClassWithArrays>(env, false).GetEnumerator() and enumerate over it.

xadupre replied:
I'll do it.

xadupre commented on test/Microsoft.ML.TestFramework/TestSparseDataView.cs:
```diff
@@ -0,0 +1,114 @@
+
+namespace Microsoft.ML.Runtime.RunTests
+{
+    public sealed partial class TestSparseDataView : TestDataViewBase
```
Copy paste without thinking :)

TomFinley commented:
Hi @xadupre thanks for addressing the comment! Clarifying note for the future, when I asked to modify the description, I meant the PR's description, not the PR's title. The description is still the little "we are excited to review your PR" built in text that we intend for people to write. I hope you don't mind, just going to change it so we can get this in since otherwise it looks good.

xadupre commented:
Hi,

Sorry for that. I did not read this part, I just checked the checkboxes. I looked into the template used by scikit-learn and the text I forgot to replace was commented out (HTML comment) and here is the discussion about it (https://github.com/scikit-learn/scikit-learn/pull/6470). That might be a good idea for us too.

Xavier



Sent from Outlook<http://aka.ms/weboutlook>

________________________________
De : Tom Finley <notifications@github.com>
Envoyé : samedi 28 juillet 2018 16:59
À : dotnet/machinelearning
Cc : Xavier Dupre; Mention
Objet : Re: [dotnet/machinelearning] Fixes #586, the creation of getters on dataviews having sparse vectors (#587)


Hi @xadupre<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fxadupre&data=02%7C01%7Cxadupre%40microsoft.com%7Cbaa687d82f064276dfc808d5f49aa905%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636683867461378704&sdata=44hUPnHMRw855yU6XeDRnBVJqt5AbS8D47Xo3hpzsMU%3D&reserved=0> thanks for addressing the comment! Clarifying note for the future, when I asked to modify the description, I meant the PR's description, not the PR's title. The description is still the little "we are excited to review your PR" built in text that we intend for people to write. I hope you don't mind, just going to change it so we can get this in since otherwise it looks good.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fdotnet%2Fmachinelearning%2Fpull%2F587%23issuecomment-408613481&data=02%7C01%7Cxadupre%40microsoft.com%7Cbaa687d82f064276dfc808d5f49aa905%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636683867461378704&sdata=hTqY1KEejcLiOP%2FdPGGrE8PnSXAvHwMMHhQ%2BE8NgQp4%3D&reserved=0>, or mute the thread<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAVaaLWE3ZD_HevTsQpdsPZCEJWuaEyMOks5uLHw3gaJpZM4VjH0j&data=02%7C01%7Cxadupre%40microsoft.com%7Cbaa687d82f064276dfc808d5f49aa905%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636683867461388711&sdata=jZxqC736MCPKl4JydC0GK8w4F2XjXeR0BxY9eGZwnGo%3D&reserved=0>.
